llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -1556,7 +1556,105 @@ class LoopVectorizationCostModel {
   getReductionPatternCost(Instruction *I, ElementCount VF, Type *VectorTy,
                           TTI::TargetCostKind CostKind) const;
 
+  using PartialReductionList = DenseMap<Instruction *, PartialReductionChain>;
+
+  PartialReductionList getPartialReductionChains() {
+    return PartialReductionChains;
+  }
+
+  bool getInstructionsPartialReduction(Instruction *I,
+                                       PartialReductionChain &Chain) const {
+    auto PairIt = PartialReductionChains.find(I);
+    if (PairIt == PartialReductionChains.end())
+      return false;
+    Chain = PairIt->second;
+    return true;
+  }
+
+  void addPartialReductionIfSupported(Instruction *Instr, ElementCount VF) {
+    Value *ExpectedPhi;
+    Value *A, *B;
+
+    using namespace llvm::PatternMatch;
+    auto Pattern =
+        m_BinOp(m_OneUse(m_BinOp(m_OneUse(m_ZExtOrSExt(m_OneUse(m_Value(A)))),
+                                 m_OneUse(m_ZExtOrSExt(m_OneUse(m_Value(B)))))),
```

Comment:
I think you don't need `m_OneUse` for the operands, you need it only for the `m_BinOp` that will be eventually lowered, correct ?
If we are doing manual unrolling, the vector will be used multiple times for multiple matrix dimensions, in that case the current pattern will not match.

Reply:
You need to make sure that the wider-than-legal vector type is never actually needed; m_OneUse on the extends was an easy way to guarantee that, but we could instead keep track of all uses of the extends to make sure they're accounted for with partial reductions.

The initial un-extended values never needed a m_OneUse, so those can indeed be removed outright.